public/app/core/components/AccessControl/AddPermission.tsx

```diff
@@ -0,0 +1,110 @@
+import React, { useEffect, useMemo, useState } from 'react';
+import { UserPicker } from 'app/core/components/Select/UserPicker';
+import { TeamPicker } from 'app/core/components/Select/TeamPicker';
+import { Button, Form, HorizontalGroup, Select } from '@grafana/ui';
+import { OrgRole } from 'app/types/acl';
+import { CloseButton } from 'app/core/components/CloseButton/CloseButton';
+import { Assignments, PermissionTarget, SetPermission } from './types';
+
+const roles = [OrgRole.Admin, OrgRole.Editor, OrgRole.Viewer];
+
+export interface Props {
+  permissions: string[];
+  assignments: Assignments;
+  canListUsers: boolean;
+  onCancel: () => void;
+  onAdd: (state: SetPermission) => void;
+}
+
+export const AddPermission = ({ permissions, assignments, canListUsers, onAdd, onCancel }: Props) => {
+  const [target, setPermissionTarget] = useState<PermissionTarget>(PermissionTarget.User);
+  const [teamId, setTeamId] = useState<number>(0);
+  const [userId, setUserId] = useState<number>(0);
+  const [builtInRole, setBuiltinRole] = useState<string>('');
+  const [permission, setPermission] = useState<string>('');
+
+  const targetOptions = useMemo(() => {
+    const options = [] as any;
+    if (assignments.users && canListUsers) {
+      options.push({ value: PermissionTarget.User, label: 'User', isDisabled: false });
+    }
+    if (assignments.teams) {
+      options.push({ value: PermissionTarget.Team, label: 'Team' });
+    }
+    if (assignments.builtInRoles) {
+      options.push({ value: PermissionTarget.BuiltInRole, label: 'Role' });
+    }
+    return options;
+  }, [assignments, canListUsers]);
+
+  useEffect(() => {
+    if (permissions.length > 0) {
+      setPermission(permissions[0]);
+    }
+  }, [permissions]);
+
+  const isValid = () => {
+    switch (target) {
+      case PermissionTarget.Team:
+        return teamId > 0;
+      case PermissionTarget.User:
+        return userId > 0;
+      case PermissionTarget.BuiltInRole:
+        return roles.some((r) => r === builtInRole);
+    }
+    return false;
```

Comment:
```suggestion
    return (
      (target === PermissionTarget.Team && teamId > 0) ||
      (target === PermissionTarget.User && userId > 0) ||
      (PermissionTarget.BuiltInRole && OrgRole.hasOwnProperty(builtInRole))
    );
```

Could also do this branchlessly if you liked. (Up to you)

Reply:
I think you suggestion is a bit cleaner